packages/client/src/blockchain/chain.ts

```diff
@@ -294,18 +294,26 @@ export class Chain {
       height: BIGINT_0,
     }
 
-    headers.latest = await this.getCanonicalHeadHeader()
+    blocks.latest = await this.getCanonicalHeadBlock()
+    blocks.finalized = (await this.getCanonicalFinalizedBlock()) ?? null
+    blocks.safe = (await this.getCanonicalSafeBlock()) ?? null
+    blocks.vm = await this.getCanonicalVmHead()
+
+    headers.latest = blocks.latest.header
     // finalized and safe are always blocks since they have to have valid execution
     // before they can be saved in chain
-    headers.finalized = (await this.getCanonicalFinalizedBlock()).header
-    headers.safe = (await this.getCanonicalSafeBlock()).header
+    if (blocks.finalized !== null) {
```

Comment:
```suggestion
   headers.finalized = blocks.finalized?.header ?? null
```
should be cleaner and simlarly for `safe`